Add addWithDefaultLabel() to ConfigurationDescriptions to allow automatically deriving the module label from the plugin name, and make first step for fillDescriptions in generic selectors

makortel opened the pull request:
Spurred by #21241  this PR presents a prototype of an fillDescriptions implementation for generic selectors (currently only for the part required to get `EventSelectorAdapter::fillDescriptions()` compile).

Summary of the design
* Add `ConfigurationDescriptions::addWithDefaultLabel()` with whcih the module label is derived from the plugin name
  * ~~Internally `@module_type` label is used to mark this special case~~
* Each component of the generic selector infrastructure that reads a parameter from `edm::ParameterSet` should implement a corresponding ~~`fillDescriptions(edm::ParameterSetDescription&)`~~ `fillPSetDescription(edm::ParameterSetDescription&)` with some default value for the parameter
* Composing components should propagate the ~~`fillDescriptions`~~ `fillPSetDescription` call to their components
* The EDModules should implement `fillDescriptions(edm::ConfigurationDescriptions&)` and call the `fillPSetDescription(edm::ParameterSetDescription&)` of their components
* The components should provide some default values for all parameters
  * ~~For module label the use of `defaultModuleLabel<T>()` from `HLTrigger/HLTcore` can lead to ridiculously long module names (e.g. `eventSelectorAdapterObjectCountEventSelectorStdVectorRecoGsfElectronStdAllocatorRecoGsfElectronPtMinSelectorMinNumberSelectorHelperCollectionFilterStdVectorRecoGsfElectronStdAllocatorRecoGsfElectronPtMinSelectorMinNumberSelector`)~~
   * ~~I guess getting a handle to the typedef names (`PtMinGsfElectronCountFilter` for above) would need macros~~
     * ~~Not included here in order to avoid going through all selector plugin definition files in case the idea doesn't fly~~
   * ~~For module label use `@module_type`~~ Insert to `ConfigurationDescriptions` with `addWithDefaultLabel()` so that ~~it~~ the label is derived automatically from the plugin name (which usually is a typedef)
* There is a mechanism to provide plugin-specific parameter default values as well (~~only for some parameters or all~~) by specializing the `fillDescriptions` function (so one has to provide values for all parameters if one want's customize even one of them)
```cpp
template <>
void TrackSelectorCountFilter::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
  edm::ParameterSetDescription desc;
  desc.add<edm::InputTag>("src", edm::InputTag("tracks"));
  desc.add<std::string>("cut", std::string("cut"));
  desc.add<unsigned int>("minNumber", 1);
  descriptions.add("trackSelectorCountFilter", desc);
}
```


I've tested in 10_0_0_pre1 that the PR compiles and ~~wf 10024.0~~ limited matrix runs. No changes are expected in monitored quantities.

makortel committed:
Remove PtMinGsfElectronCountFilter as duplicate

This file is a complete duplicate of
CommonTools/RecoAlgos/plugins/PtMinGsfElectronCountFilter.cc
and is not used anywhere in CMSSW, so I remove it.

makortel committed:
Prototype of fillDescpritions for generic selectors

With a generic mechanism to provide default values to the
fillDescriptions. By default the module labels can get ridiculously
long though.

cmsbuild commented:
The code-checks are being triggered in jenkins.

makortel commented on CommonTools/UtilAlgos/interface/EventSelectorAdapter.h:
```diff
@@ -17,6 +17,14 @@
+template <typename T>
+struct FillDescriptionTraits {
+  static std::string moduleLabel() { return defaultModuleLabel<T>(); }
+};
```
If we take this route, obviously this definition must go to some more "central" header (and the class template to be renamed and/or put in a namespace).

slava77 replied:
I think that this implementation is not very useful because it requires knowledge of the mangled name, while in reality it's the registered module plugin name that is the most relevant default.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21430/2146

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

fwyzard commented:
One minor worry: calling `fillDescriptions` both the usual static method and the new one for the components might be confusing (I know it would have been confusing for me when I first started playing with these).
Can anyone suggest a slightly different name ?
I would propose `fillPSetDescription` or `fillDescriptionFragment` or `fillDescriptionPart`, but I don't really like any of them :(

cmsbuild commented:
-1

Tested at: 5055f5174731e3a33813d268be49c6bf63db03c3


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/5780cd48e28d35f8052219e1dd10cb8f9fae4c43
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24608/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24608/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24608/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
1000.0 step3
<pre>runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step3_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log

</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/5780cd48e28d35f8052219e1dd10cb8f9fae4c43
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24608/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24608/git-merge-result

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

slava77 commented:
the error is from -s SKIM:all
```
   [1] Validating configuration of module: class=CandViewCountFilter label='onia2MuMuPATCounter'
Exception Message:
Illegal parameter found in configuration.  The parameter is named:
 'filter'
You could be trying to use a parameter name that is not
allowed for this plugin or it could be misspelled.

```

makortel commented:
Repeating my question in https://github.com/cms-sw/cmssw/pull/21443#discussion_r152919707 here (for completeness):

Indeed this PR and #21443 /Slava's solution can coexist. If we then take the route of "coexistence", do we need the fillDescriptions customization mechanism here (that allows overriding individual parameters), or is it sufficient to say that if user wants to customize anything (including the module label) (s)he must specify the full fillDescriptions?

I'm asking because removing the customization mechanism would make it simpler (e.g. no need for SFINAE checks).

slava77 commented on CommonTools/UtilAlgos/interface/EventSelectorAdapter.h:
```diff
@@ -17,6 +17,14 @@
 #include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "HLTrigger/HLTcore/interface/defaultModuleLabel.h"
```
if needed in this package, this file will need to be moved to some place in FWCore

slava77 commented:
> or is it sufficient to say that if user wants to customize anything (including the module label) (s)he must specify the full fillDescriptions?

I think that ability to specify a module label is important if there is no other way to make the default module name to be derived from the module plugin name.
I don't have a strong feeling about all other parameters.

makortel commented:
> > or is it sufficient to say that if user wants to customize anything (including the module label) (s)he must specify the full fillDescriptions?
> 
> I think that ability to specify a module label is important if there is no other way to make the default module name to be derived from the module plugin name.

I'm not aware of any other way to get a handle to the plugin name (typedef/alias) than a macro (would be happy to learn if there is). I don't know if `DEFINE_FWK_MODULE()` would already provide something that could be used (I'd guess no). At minimum, to minimize copy paste, a new macro can be created (e.g. `SELECTOR_MODULE_LABEL()` or something).

slava77 commented:
@makortel 
if we go with this PR, perhaps it's time to drop the "[RFC]"

makortel commented:
(@slava77 let me go with one round of questions before dropping the RFC)

> > > or is it sufficient to say that if user wants to customize anything (including the module label) (s)he must specify the full fillDescriptions?
> > 
> > 
> > I think that ability to specify a module label is important if there is no other way to make the default module name to be derived from the module plugin name.
> 
> I'm not aware of any other way to get a handle to the plugin name (typedef/alias) than a macro (would be happy to learn if there is). I don't know if `DEFINE_FWK_MODULE()` would already provide something that could be used (I'd guess no). At minimum, to minimize copy paste, a new macro can be created (e.g. `SELECTOR_MODULE_LABEL()` or something).

I started to wonder if the "default module label" concept would be useful-enough to actually enhance `DEFINE_FWK_MODULE()` macro so that it does the necessary "magic" to deliver the label from the module C++ type to a function that could be easily called from `fillDescriptions()`. More concretely
* add a call of another macro to `DEFINE_FWK_MODULE` with the "module type" as an argument
* the other macro specializes a "traits class template" (e.g. `edm::DefaultModuleLabel`) to return the "module type" name with first letter converted to lower case (as this is already a common convention)
   * or the logic from
https://github.com/cms-sw/cmssw/blob/17e3808b0f6af1864c8cf37d26ad8209d03f5bd9/HLTrigger/HLTcore/interface/defaultModuleLabel.h#L47-L59
* the "new function" would then just return the label from the "traits template"

This way the default label from the "module type" (typedef, not full type name demangled) would be automatically available everywhere (but not forced), including generic modules outside of these selectors. 

I know this is going somewhat beyond the original scope of the prototype, but thought to ask anway. @Dr15Jones  what do you think?

slava77 commented:
On 11/27/17 6:57 AM, Matti Kortelainen wrote:
> (@slava77 <https://github.com/slava77> let me go with one round of
> questions before dropping the RFC)
> 
>             or is it sufficient to say that if user wants to customize
>             anything (including the module label) (s)he must specify the
>             full fillDescriptions?
> 
>         I think that ability to specify a module label is important if
>         there is no other way to make the default module name to be
>         derived from the module plugin name.
> 
>     I'm not aware of any other way to get a handle to the plugin name
>     (typedef/alias) than a macro (would be happy to learn if there is).
>     I don't know if |DEFINE_FWK_MODULE()| would already provide
>     something that could be used (I'd guess no). At minimum, to minimize
>     copy paste, a new macro can be created (e.g.
>     |SELECTOR_MODULE_LABEL()| or something).
> 
> I started to wonder if the "default module label" concept would be
> useful-enough to actually enhance |DEFINE_FWK_MODULE()| macro so that it
> does the necessary "magic" to deliver the label from the module C++ type
> to a function that could be easily called from |fillDescriptions()|.
> More concretely
> 
>   * add a call of another macro to |DEFINE_FWK_MODULE| with the "module
>     type" as an argument
>   * the other macro specializes a "traits class template" (e.g.
>     |edm::DefaultModuleLabel|) to return the "module type" name with
>     first letter converted to lower case (as this is already a common
>     convention)

this should probably also do something about the namespace colons, which
are legal and are present in some modules, e.g. l1t::BXVectorInputProducer.
My guess is that it is not good to have a double-colon in python module
names.


>       o or the logic from
>         https://github.com/cms-sw/cmssw/blob/17e3808b0f6af1864c8cf37d26ad8209d03f5bd9/HLTrigger/HLTcore/interface/defaultModuleLabel.h#L47-L59
>   * the "new function" would then just return the label from the "traits
>     template"
> 
> This way the default label from the "module type" (typedef, not full
> type name demangled) would be automatically available everywhere (but
> not forced), including generic modules outside of these selectors.
> 
> I know this is going somewhat beyond the original scope of the
> prototype, but thought to ask anway. @Dr15Jones
> <https://github.com/dr15jones> what do you think?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/21430#issuecomment-347207179>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbh7EL-HRT0z5Sdr40DMLbvaycMJeks5s6s3ZgaJpZM4QnZVX>.
>

Dr15Jones commented:
>this should probably also do something about the namespace colons, which
are legal and are present in some modules, e.g. l1t::BXVectorInputProducer.
My guess is that it is not good to have a double-colon in python module
names.

I personally feel it is much better to have the namespaces than to not have the namespaces in the module type specification.

makortel commented:
> > this should probably also do something about the namespace colons, which
> > are legal and are present in some modules, e.g. l1t::BXVectorInputProducer.
> > My guess is that it is not good to have a double-colon in python module
> > names.

Good point (didn't realize that).

> I personally feel it is much better to have the namespaces than to not have the namespaces in the module type specification.

IIUC Slava was against the colons. So for `l1t::BXVectorInputProducer` the default module label should be `l1tBXVectorInputProducer`?

slava77 commented:
On 11/27/17 7:39 AM, Matti Kortelainen wrote:
>         this should probably also do something about the namespace
>         colons, which
>         are legal and are present in some modules, e.g.
>         l1t::BXVectorInputProducer.
>         My guess is that it is not good to have a double-colon in python
>         module
>         names.
> 
> Good point (didn't realize that).
> 
>     I personally feel it is much better to have the namespaces than to
>     not have the namespaces in the module type specification.
> 
> IIUC Slava was against the colons. So for |l1t::BXVectorInputProducer|
> the default module label should be |l1tBXVectorInputProducer|?

right.
It's fine to keep the plugin names as they are, but the python module
instances will have to drop the colons. IIRC, a colon is not a legal
character in the EDM product name.


> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/21430#issuecomment-347220742>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbqLHSJr-3DCNNKJscOPib0cccSNxks5s6te6gaJpZM4QnZVX>.
>

makortel committed:
Simplify

Remove per-parameter customization options to simplify. To customize
parameters the user is expected to specialize the fillDescriptions()
member function of the EDModule. The default module label will be
improved in subsequent commits.

makortel committed:
fillDescriptions -> fillPSetDescription for the helper functions

makortel committed:
Add edm::defaultModuleLabel() to convert plugin name to a default module label

makortel commented:
Apparently doing any kind of template specialization within `DEFINE_FWK_MODULE()` is a no-go because of constructs like these
https://github.com/cms-sw/cmssw/blob/efba733c40b1f49ad9f83dc706e40349a0393c13/CommonTools/RecoAlgos/plugins/RecoTrackSelector.cc#L16-L22
as templates must be specialized in the same namespace where they were defined (more or less).

I have to think a bit if there is anything better (from ease-of-use point of view) that could be done besides calling another macro explicitly (that must be placed in the global namespace).

makortel commented:
(In case anyone else would have bright ideas, the challenge is getting the module "typedef name" as string from the full type. (Class) template specialization would have been an easy way. I could also think of a global string->string mapping filled with a similar mechanism as plugin factories, but it doesn't "feel right".)

Dr15Jones commented:
The `typedef` name used to specify the module in the configuration is available in the ParameterSet itself under the parameter `"@module_type"`.

makortel commented:
@Dr15Jones Ok, but is it also available from `ConfigurationDescriptions`? (by quick look I'd say "no")

Or some other way for class `T` within `T::fillDescriptions()`?

makortel commented:
> Ok, but is it also available from `ConfigurationDescriptions`? (by quick look I'd say "no")

After a bit longer look I'd indeed say "no" presently. But in all places where `ConfigurationDescriptions` is constructed the `typedef` name is otherwise available (either there is a `ParameterSet` from which to ask for `@module_type`, or the plugin name is used to construct the module). So we could easily pass it to `ConfigurationDescriptions` constructor.

If that is acceptable, I'm wondering whether the "plugin name" should be exposed from `ConfigurationDescriptions` or not. I feel code like
```cpp
descriptions.add(defaultModuleLabel(descriptions.getPluginName()), ...);
// or
descriptions.add(descriptions.defaultModuleLabel(), ...);
```
would look a bit silly. So maybe something like
```
descriptions.add(ConfigurationDescriptions::defaultLabel, ...);
// or
descriptions.addWithDefaultLabel(...);
```
?

Dr15Jones commented:
Let's take a step back. Why are you trying to get a default module label? If it is to generate the _cfi.py file, then I think it would be better to modify `ConfigurationDescriptions::writeCfis` which is passed in the `pluginName` which corresponds to the typedef. One could use a special label name, say "@moduleType", which that function looks for and would then substitute with a name coming from `pluginName`.

makortel commented:
That would certainly be quite simple. I'll try that (`descriptions.add("@module_type", ...)`) for the next version. Thanks!

makortel committed:
Add ConfigurationDescriptions::addWithDefaultLabel() which derives the default module label from the plugin name

Requires giving plugin name to the ConfigurationDescriptions constructor

makortel committed:
Use addWithDefaultLabel in EventSelectorAdapter

makortel committed:
CandViewCountFilter does not have 'filter' parameter

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21430/2247

cmsbuild commented:
Pull request #21430 was updated. @perrotta, @smuzaffar, @prebello, @Dr15Jones, @fabozzi, @cmsbuild, @slava77, @GurpreetSinghChahal, @monttj can you please check and sign again.

makortel commented:
Did the following updates
* removed the ability to customize individual parameters (as likely unnecessary complexity)
  * to override defaults for some module user has to specialize the full `fillDescriptions()` method
* Renamed `fillDescriptions()` to `fillPSetDescriptions()` for those taking `edm::ParameterSetDescription`  parameter (following suggestion by @fwyzard in https://github.com/cms-sw/cmssw/pull/21430#issuecomment-346363769, picked the least bad option to my taste in absence of better ideas)
* `ConfigurationDescriptions::writeCfiForLabel()` constructs the label from plugin name if the given label is `@module_type`
   * The plugin name -> default label conversion logic is implemented in a separate function
* Use `@module_type` for `EventSelectorAdaptor`
  * I'd prefer to leave the remaining selector EDModule templates to subsequent PR(s)
* Remove `filter` parameter for `CandViewCountFilter` in the BPH skim (as the parameter does not exist in the EDModule)

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

makortel commented on FWCore/ParameterSet/src/ConfigurationDescriptions.cc:
```diff
@@ -176,18 +177,19 @@ namespace edm {
 
-    std::string cfi_filename;
-    if (0 == strcmp(baseType.c_str(),kSource)) {
-      cfi_filename = pluginName + "_cfi.py";
```
I took the liberty of simplifying a bit the logic here (and below) since for services the `labelAndDesc.first == pluginName` is guaranteed already by the check above.

makortel replied:
Argh, it was `kSource`, not `kService`.

makortel commented on CommonTools/RecoAlgos/plugins/PtMinPixelMatchGsfElectronCountFilter.cc:
```diff
@@ -1,20 +0,0 @@
-typedef ObjectCountFilter<
-          reco::GsfElectronCollection, 
-          PtMinSelector
-        >::type PtMinPixelMatchGsfElectronCountFilter;
```
In principle with the current implementation of the default label this module could co-exist with `PtMinGsfElectronCountFilter.cc`, but as noted in the commit message the definition is exactly the same
https://github.com/cms-sw/cmssw/blob/e775c2b6b822417a83d14d30a545fc5628003c15/CommonTools/RecoAlgos/plugins/PtMinGsfElectronCountFilter.cc#L15-L18
and there are no uses within CMSSW. My preference would be to keep the deletion commit.

cmsbuild commented:
-1

Tested at: 180c37bf6815a2866e60a347c31d6189f729daba


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/8eb7f2c0e3f7fd3f5e23bdc9f7435fee41d8402d
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24754/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24754/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24754/summary.html

I found follow errors while testing this PR

**Failed tests:  Build**

* **Build**:

I found an error when building: 
 
 <pre>>> Compiling python modules src/Validation/RecoParticleFlow/python
>> Done generating edm plugin poisoned information
>> All python modules compiled
>> Plugins of all types refreshed.
gmake[1]: Leaving directory '/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-11-29-1100'
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 1
</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/8eb7f2c0e3f7fd3f5e23bdc9f7435fee41d8402d
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24754/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24754/git-merge-result

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

makortel commented:
Hmm, didn't catch that in my local test...

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21430/2248

cmsbuild commented:
Pull request #21430 was updated. @perrotta, @smuzaffar, @prebello, @Dr15Jones, @fabozzi, @cmsbuild, @slava77, @GurpreetSinghChahal, @monttj can you please check and sign again.

makortel commented:
@cmsbuild, please test

Let's try again

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 0737322ddf6db45f85715bc44a737038c153fbde
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24755/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/8eb7f2c0e3f7fd3f5e23bdc9f7435fee41d8402d
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24755/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24755/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24755/summary.html

There are some workflows for which there are errors in the baseline:
 10824.0 step 5
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2788080
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2787901
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.570000000138 KiB( 23 files compared)
 - Checked 111 log files, 8 edm output root files, 27 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/21430	0737322
- implementation is in line with the description (it was updated to be in sync with the final version)
- jenkins tests pass and comparisons with baseline show no differences
- local build shows that the new generated config files show up as expected 
e.g. 
```
CommonTools/RecoAlgos/vertexCountFilter_cfi.py
DPGAnalysis/Skims/pixelCountFilter_cfi.py
PhysicsTools/PatAlgos/patCandViewCountFilter_cfi.py
```
- expanded configs for the short matrix workflows are unchanged except for the removal of the superfluous ``` filter = cms.bool(True)``` in onia2MuMuPATCounter

fwyzard commented on CommonTools/UtilAlgos/interface/EventSelectorAdapter.h:
```diff
@@ -30,6 +32,12 @@ class EventSelectorAdapter : public edm::global::EDFilter<>
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+    edm::ParameterSetDescription desc;
+    T::fillPSetDescription(desc);
+    descriptions.add("@module_type", desc);
```
if I understand correctly, when run for a module of type "MyModule" this will result in a `MyModule_cfi.py` file being generated, with the definition of `MyModule = cms.EDProducer("MyModule", ...)`.  
According to some guidelines we used years back, and the more common use case, this should rather go into a `myModule_cfi.py` with a definition like `myModule = cms.EDProducer("MyModule", ...)`.  

Would you consider making the first letter lowercase ?

slava77 commented:
On 12/1/17 1:53 PM, Andrea Bocci wrote:
> *@fwyzard* requested changes on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In CommonTools/UtilAlgos/interface/EventSelectorAdapter.h
> <https://github.com/cms-sw/cmssw/pull/21430#discussion_r154457535>:
> 
>> @@ -30,6 +32,12 @@ class EventSelectorAdapter : public edm::global::EDFilter<>
>    // destructor
>    ~EventSelectorAdapter() override {}
>  
> +  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
> +    edm::ParameterSetDescription desc;
> +    T::fillPSetDescription(desc);
> +    descriptions.add("@module_type", desc);
> 
> if I understand correctly, when run for a module of type "MyModule" this
> will result in a |MyModule_cfi.py| file being generated, with the
> definition of |MyModule = cms.EDProducer("MyModule", ...)|.
> According to some guidelines we used years back, and the more common use
> case, this should rather go into a |myModule_cfi.py| with a definition
> like |myModule = cms.EDProducer("MyModule", ...)|.
> 
> Would you consider making the first letter lowercase ?

It is the case already
see FWCore/ParameterSet/src/ConfigurationDescriptions.cc   and case
changes in
FWCore/ParameterSet/interface/defaultModuleLabel.h


An example content for CommonTools/CandAlgos/candViewCountFilter_cfi.py
is
candViewCountFilter = cms.EDFilter('CandViewCountFilter',
	  src = cms.InputTag(''),
	  minNumber = cms.uint32(0)
	)

> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/21430#pullrequestreview-80631738>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbj4FYgz-HXUMTzRPKd5Atjop7oTKks5s8HVOgaJpZM4QnZVX>.
>

fwyzard commented:
Ah, I see, it doesn't expand `@module_type` to the module type, it uses it as a token to mean "generate the default label from the C++ type".

Then, rather than explicitly using `"@module_type"`, I would suggest to use some mnemonic like `defaultLabel` (could be `edm::ParameterSetDescription::defaulLabel` or `edm::defaulLabel`),e.g. `descriptions.add(edm::defaultLabel, desc)` .

Even better, I would suggest to use `descriptions.addDefaultLabel(desc)`.

fwyzard commented:
There should also be a special case for a `Service`, where the label needs to be the same as the plugin type, without any case adjustments.

fwyzard commented on FWCore/ParameterSet/interface/defaultModuleLabel.h:
```diff
@@ -0,0 +1,30 @@
+
+    // if the label is all uppercase, change it to all lowercase
+    // if the label starts with more than one uppercase letter, change n-1 to lowercase
+    // otherwise, change the first letter to lowercase
```
I know that the logic is trivial, still I would have appreciated something along "taken from HLTrigger/HLTcore/interface/defaultModuleLabel.h" in a C++ comment or in the commit message :-(

makortel replied:
Good point, I added a comment noting the origin.

fwyzard replied:
Thanks :-)

makortel committed:
Give credit where it belongs

cmsbuild commented:
The code-checks are being triggered in jenkins.

makortel commented:
Addressed @fwyzard's comments (thanks!).

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21430/2409

cmsbuild commented:
Pull request #21430 was updated. @perrotta, @smuzaffar, @prebello, @Dr15Jones, @fabozzi, @cmsbuild, @slava77, @GurpreetSinghChahal, @monttj can you please check and sign again.

makortel commented:
@fwyzard 
> Even better, I would suggest to use `descriptions.addDefaultLabel(desc)`

Thanks for reminding, I added `addWithDefaultLabel(desc)`. The `@module_type` is still used internally.

> There should also be a special case for a Service, where the label needs to be the same as the plugin type, without any case adjustments.

Good point, fixed.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
@wddgit please review

cmsbuild commented:
-1

Tested at: ef1c1aedabee1d1a1af2566ffe190c9db28a5fd0


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/62081f815696096b571931b8fefc7931a0bf18b3
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24923/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24923/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24923/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testTauEmbeddingProducers had ERRORS



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/62081f815696096b571931b8fefc7931a0bf18b3
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24923/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24923/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24923/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-21430/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017+NANOMERGE

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2835085
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2834906
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.60000000002 KiB( 23 files compared)
 - Checked 114 log files, 10 edm output root files, 27 DQM output files

wddgit commented:
@makortel 

Mostly it looks good to me, but I noticed a few problems. (I did not test this and it is possible I am not remembering correctly how this works since it has been a long time since I was into this code).

1. The system allows multiple configuration descriptions for a particular plugin. When validating it selects a description based on the label with the following criteria in this order: exact label match, then explicit default description if there is one, then first description, then if there are no descriptions it skips validation. It looks like the special label @module_type is going into the vector where the lookups are done and it will try to match it with the actual module label. This will work OK if there is only one description and in the way you are using this function there is only one, but the new function is in the interface and there is nothing to prevent someone else from using it in other ways. It will also work OK if it is the first description and there is no default description. But in others cases, it looks like the wrong description will be selected for validation. One possible solution would be to only allow one description and no explicit default description in this new case.  You could throw an exception if another description were added when there was already one of this special type. Another solution would be to require it to be first and not allow a default in this case. Another would be to derive the label earlier and get it into the vector where the searches are done. I am not sure how easy this is. There might be some other better way to fix this. Here is where this validation logic is located.

http://cmslxr.fnal.gov/source/FWCore/ParameterSet/src/ConfigurationDescriptions.cc#0128

2. It looks like the label will be wrong in the case of a source. You could check by looking at the generated cfi for PoolSource here:
  cfipython/slc6_amd64_gcc630/IOPool/Input/PoolSource_cfi.py
It should look like this:
  source = cms.Source('PoolSource',

3. It is not as important but the exception message in the case of duplicate cfi filenames for a source will say the label is the plugin name, which is erroneous.

Other than that it looks good to me.

slava77 commented:
> ---> test testTauEmbeddingProducers had ERRORS

the crash is likely unrelated to this PR.
The errors are a bit "funny"

```
----- Begin Fatal Exception 04-Dec-2017 00:00:01 CET-----------------------
An exception of category 'ExceededResourceRSS' occurred while
   [0] Calling EventProcessor::runToCompletion (which does almost everything after beginJob and before endJob)
Exception Message:
Exceeded maximum allowed RSS of 0.001 GB (VSize is 0.07575)
```

and another one
```
----- Begin Fatal Exception 04-Dec-2017 00:00:04 CET-----------------------
An exception of category 'ExceededResourceTime' occurred while
   [0] Calling EventProcessor::runToCompletion (which does almost everything after beginJob and before endJob)
Exception Message:
Exceeded maximum allowed time of 2.77778e-06 hours
```

Since it's the first time I see similar errors, what could have caused them and how to fix if reappear again?

Dr15Jones commented:
@slava77 what you copied was not the error. Those are just expected results for successful tests in the framework. In order to find the error you need to search the file for
`---> test testTauEmbeddingProducers had ERRORS`

Right before that line we get
```
----- Begin Fatal Exception 04-Dec-2017 00:02:47 CET-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=CandViewCountFilter label='ZmumuCandidatesFilter'
Exception Message:
Illegal parameter found in configuration.  The parameter is named:
 'filter'
You could be trying to use a parameter name that is not
allowed for this plugin or it could be misspelled.
----- End Fatal Exception -------------------------------------------------
```

which does look related to this pull request.

slava77 commented:
On 12/4/17 11:31 AM, Chris Jones wrote:
> |Illegal parameter found in configuration. The parameter is named:
> 'filter' You could be trying to use a parameter name that is not allowed
> for this plugin or it could be misspelled. ----- End Fatal Exception
> ------------------------------------------------- |
> 
> which does look related to this pull request.
> 

this makes more sense.
Thanks.

I got confused by somewhat out-of-order test result printouts

makortel committed:
Remove remaining filter argument to CandViewCountFilter

makortel commented:
@wddgit Thanks for the comments.

Regarding your point 1, I think the simplest option would be to follow
> Another would be to derive the label earlier and get it into the vector where the searches are done."

by passing the plugin name to the constructor of `ConfigurationDescriptions` (which I thought of already in https://github.com/cms-sw/cmssw/pull/21430#issuecomment-347661331). Then `ConfigurationDescriptions::addWithDefaultLabel()` could transform the plugin name to the default label and push the `ParameterSetDescription` with that label to the system (without playing with special values like `@module_type`). This option would also automatically solve points 2 and 3 since I could just revert all my changes to `ConfigurationDescriptions::writeCfiForLabel()`.

Alternatively we could pass the plugin name to the `ConfigurationDescriptions::validate()`, but in practice that would require changes to all places constructing `ConfigurationDescriptions` (except `edmWriteConfigs`) so this option would be only more complex than what I wrote above.

I'm not very keen to start restricting the interface since there is a simple-looking alternative and any checks can be done only at run-time as you wrote.

cmsbuild commented:
The code-checks are being triggered in jenkins.

makortel commented:
In the meantime I fixed the failing unit test.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21430/2433

cmsbuild commented:
Pull request #21430 was updated. @perrotta, @smuzaffar, @prebello, @Dr15Jones, @civanch, @mdhildreth, @fabozzi, @cmsbuild, @slava77, @GurpreetSinghChahal, @monttj can you please check and sign again.

makortel commented:
@wddgit @Dr15Jones 

Additional deficiency of the PR as it stands is that `edmPluginHelp` will give `module label: @module_type`, which, I believe, is not very helpful. I think also this issue would be automatically solved with what I wrote in https://github.com/cms-sw/cmssw/pull/21430#issuecomment-349309867.

makortel committed:
Simplify the interface of ConfigurationDescriptions::writeCfis()

These parameters are already member variables.

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 1182adc7e2b244213f136e39e15e22137381d4fe
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24949/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/78a797a46a83721d8571bcf954088127f010a183
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24949/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24949/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21430/2459

cmsbuild commented:
Pull request #21430 was updated. @perrotta, @smuzaffar, @prebello, @Dr15Jones, @civanch, @mdhildreth, @fabozzi, @cmsbuild, @slava77, @GurpreetSinghChahal, @monttj can you please check and sign again.

makortel commented:
@wddgit @Dr15Jones 

I decided to go a head (to demonstrate if nothing else) with passing the plugin name to `ConfigurationDescriptions` constructor and using that in `ConfigurationDescriptions::addWithDefaultLabel()`. Could you please take a look and comment if there are further issues (or if something could be done in a simpler way)?

I did the changes as incremental commits by reverting most of commits I did for the `@module_type` approach. If the current approach gets accepted, I'd like to squash the commits for a cleaner history.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

wddgit commented:
@makortel The changes look good to me now. That looks like the best way to resolve the problems. You could add to the Core unit test. Add something like:

edm::ParameterSetDescription iDesc3;
iDesc3.add<int>("p_int", 3);
descriptions.addWithDefaultLabel(iDesc3);

to file FWCore/Integration/test/ProducerWithPSetDesc.cc near the end of the file. This gets used in the test run by this file: http://cmslxr.fnal.gov/source/FWCore/Integration/test/run_ParameterSet.sh. This will require you to update at least 2 reference files. (Run "scram b runtests" and copy the files from the ../tmp/slc6_amd64_gcc630/testProducerWithPsetDesc_briefdoc.txt and testProducerWithPsetDesc_doc.txt
to the FWCore/Integration/test/unit_test_outputs area if the differences are what you expect). You also could add a new reference file and a diff test for the new cfi file that should be generated similar to the existing one.

cmsbuild commented:
+1
Tested at: 7f94ffe9bda7b3b734a70966048034a5cc97d4ec
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24966/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/0179dc54fdff21005908765fa22a5a152179e983
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24966/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24966/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24966/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-21430/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2835085
 - DQMHistoTests: Total failures: 21
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2834886
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.18999999994 KiB( 22 files compared)
 - Checked 113 log files, 9 edm output root files, 27 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/21430	7f94ffe
- somewhat technical updates from reco perspective since the last signoff https://github.com/cms-sw/cmssw/compare/0737322ddf6db45f85715bc44a737038c153fbde...7f94ffe9bda7b3b734a70966048034a5cc97d4ec
- jenkins tests pass and comparisons with baseline show no relevant differences (ignoring occasional LogError/warning in the PU workflow and a bunch of L1 tau plots which have some random small differences visible in some other PR tests and also seen in the past)
- local tests checking the cfipython files and expanded configs from the short matrix workflows show that since the last reco signoff there were no changes in created configurations.

Dr15Jones commented:
+1

makortel commented:
Now that @wddgit and @Dr15Jones have approved the implementation, let me repeat that I'd prefer to squash some of the commits to remove the going-back-and-forth from the git history. I'm also working for the unit test @wddgit suggested in https://github.com/cms-sw/cmssw/pull/21430#issuecomment-349690538.

makortel committed:
Add unit test for addWithDefaultLabel()

cmsbuild commented:
The code-checks are being triggered in jenkins.

makortel commented:
Did the squashing and added the unit test. The state at new commit d6b5974 should be exactly the same as in old commit 7f94ffe9bda. I'm done now (unless something new surfaces).

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21430/2492

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #21430 was updated. @perrotta, @smuzaffar, @prebello, @Dr15Jones, @civanch, @mdhildreth, @fabozzi, @cmsbuild, @slava77, @GurpreetSinghChahal, @monttj can you please check and sign again.

Dr15Jones commented:
+1

cmsbuild commented:
+1
Tested at: 4a59669ca7b9e5e281c9147ff570286f664a9fa0
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24994/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/2699c4c446b0a1dc00b007abba3ee486763ccc27
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24994/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24994/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21430/24994/summary.html

There are some workflows for which there are errors in the baseline:
 10824.0 step 5
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-21430/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2798638
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2798459
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.879999999866 KiB( 23 files compared)
 - Checked 113 log files, 9 edm output root files, 27 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/21430	4a59669
- technical update since the last signoff: a test with ```descriptions.addWithDefaultLabel``` is added in FWCore/Integration/test/ProducerWithPSetDesc.cc, no other changes
- jenkins tests pass

civanch commented:
+1

davidlange6 commented:
merge
